Add package validation command

landerweit-phytec opened the pull request:
Add a new command for package validation `validate`.
The goal of package validation is to make sure a package can be written to a device without writing to the device. Additionally, remove most validation steps from the `install` command to reduce the time i takes to write to the device.
Open tasks:
- [x] Check if URI points into archive
- [x] Size calculation for uncompressed archives
- [x] Option to use `validate` without passing a device (maybe use a loop device and pass eMMC size as argument)
- [x] Ensure the partitions fit the device
- [x] Test validate command
- [x] Add commit messages

landerweit-phytec committed:
src: Make input files independent from emmc

Separate input files from flash specific information as they only
contain general information like filename, checksums and size on target.

Signed-off-by: Leonard Anderweit <l.anderweit@phytec.de>

landerweit-phytec committed:
src: Add config validation function

Add a new function for config validation. For now only verifies
checksums.

Signed-off-by: Leonard Anderweit <l.anderweit@phytec.de>

landerweit-phytec committed:
src: Remove checksum verification from install command

Checksum verification is done with the validate command. So remove it
from the install command to save time when writing to a device.
Update the docs accordingly.

Signed-off-by: Leonard Anderweit <l.anderweit@phytec.de>

landerweit-phytec committed:
Check file size and raw overlap only in validate

These steps are not necessary for package installation so execute them
only when validating a package.
Also, update the tests accordingly.

Signed-off-by: Leonard Anderweit <l.anderweit@phytec.de>

landerweit-phytec committed:
tests: Add unit test for pu-input

Signed-off-by: Leonard Anderweit <l.anderweit@phytec.de>

landerweit-phytec committed:
src: emmc: Check if input files fit into partition

Add an additional validation step to check if the input files of each
partition fit into the corresponding partition.

Signed-off-by: Leonard Anderweit <l.anderweit@phytec.de>

landerweit-phytec committed:
src: emmc: Get absolute path and file size for all input files centrally

Add a list for all input files regardless of their destination. Use
it to get the absolute path and size of all input files at once after
config parsing instead of during writing.

Signed-off-by: Leonard Anderweit <l.anderweit@phytec.de>

landerweit-phytec committed:
github: workflows: system-tests: Add package validation step

Signed-off-by: Leonard Anderweit <l.anderweit@phytec.de>

landerweit-phytec committed:
src: utils: Add function to parse size strings

Add a new function pu_parse_size() which accepts a size string and
returns the size in bytes. Accepts strings with the same format as
ped_parse_unit, but only with absolute units.
Also, add a corresponding unit test.

Signed-off-by: Leonard Anderweit <l.anderweit@phytec.de>

landerweit-phytec committed:
doc: usage: Add validate command

Signed-off-by: Leonard Anderweit <l.anderweit@phytec.de>

landerweit-phytec committed:
src: emmc: Check if partitions fit into device

Check if the fixed size partitions fit into the device. This does not
check expanding partitions.

Signed-off-by: Leonard Anderweit <l.anderweit@phytec.de>

landerweit-phytec committed:
src: utils: Add function to get the stdout output of a command

Add a new function pu_spwan_command_line_sync_result() which returns the
output of stdout.
Also, add a corresponding unit test.

Signed-off-by: Leonard Anderweit <l.anderweit@phytec.de>

landerweit-phytec committed:
tests: emmc: Add unit tests for validation

Signed-off-by: Leonard Anderweit <l.anderweit@phytec.de>

mschwan-phytec commented on src/pu-utils.c:
```diff
@@ -496,3 +506,66 @@ pu_str_pre_remove(gchar *string,
 }
+
+gboolean
+pu_parse_size(const gchar *string,
```
Instead of reinventing the wheel, have you considered using `ped_unit_parse()` from libparted?

landerweit-phytec replied:
Unfortunately `ped_unit_parse()` can not be used here as it requires a device of some sort, which is not present when `pu_parse_size()` is used.

landerweit-phytec replied:
Yes, it can be used with loop devices. This is the main reason I introduced loop devices here.
The problem here is that we need some way to determine the initial size of the loop device before creating it. And that is where `pu_parse_size()` is used.

mschwan-phytec commented on src/pu-emmc.c:
```diff
@@ -164,6 +158,112 @@ emmc_create_partition(PuEmmc *self,
+                 "skip-checksums", &skip_checksums,
+                 NULL);
+
+    if (!skip_checksums)
+        for (GList *i = self->input_files; i != NULL; i = i->next) {
+            PuInput *input = i->data;
+            if (!pu_validate_file(input, error))
+                return FALSE;
+        }
```
Use curly braces to enclose the if-statement content. It's currently a little confusing to read.

landerweit-phytec replied:
Fixed

mschwan-phytec reviewed (changes requested):
A first glimpse on your PR...

mschwan-phytec commented:
You may want to consider squashing some commit fixing earlier stuff of your PR, e.g.:
```
Check file size and raw overlap only in validate
```
There is code in `src/pu-emmc.c:pu_emmc_new` that gets added earlier and then removed. Maybe you can squash/rearrange your commits, so changes are correct in the first place.

mschwan-phytec commented on doc/usage.rst:
```diff
@@ -50,6 +48,13 @@ show [OPTION…] *PACKAGE*
+validate [OPTION…] *PACKAGE* *SIZE*
+   Validate a partup PACKAGE
+
+   -s, --skip-checksums         Skip checksum verification for all input files
```
The option `--size` is missing here.

landerweit-phytec replied:
Fixed

mschwan-phytec commented on doc/usage.rst:
```diff
@@ -50,6 +48,13 @@ show [OPTION…] *PACKAGE*
+validate [OPTION…] *PACKAGE* *SIZE*
+   Validate a partup PACKAGE
+
+   -s, --skip-checksums         Skip checksum verification for all input files
+   -w, --write-file             Install the package to a file as an additional validation step
+   -k, --keep-file-as=FILENAME  Keep the file. Implies --write-file
```
From a user perspective I think it is not clear, what you mean exactly with "file" in the options `--write-file` and `--keep-file-as` as the term collides with the input files.

mschwan-phytec commented on src/pu-main.c:
```diff
@@ -220,13 +363,30 @@ static GOptionEntry option_entries_version[] = {
+    { "skip-checksums", 's', G_OPTION_FLAG_NONE, G_OPTION_ARG_NONE,
+        &arg_validate_skip_checksums, "Skip checksum verification for all input files", NULL },
+    { "size", 0, G_OPTION_FLAG_NONE, G_OPTION_ARG_STRING,
+        &arg_validate_file_size, "Size of loop device. Default is 1G", "SIZE" },
```
Again, a user may not be familiar with loop devices, so this description is not easy to understand without looking at the inner workings of partup.

mschwan-phytec commented on src/pu-main.c:
```diff
@@ -220,13 +363,30 @@ static GOptionEntry option_entries_version[] = {
+    { "keep-file-as", 'k', G_OPTION_FLAG_NONE, G_OPTION_ARG_FILENAME,
+        &arg_validate_keep_file, "Keep the file. Implies --write-file", "FILENAME" },
+    { G_OPTION_REMAINING, 0, G_OPTION_FLAG_NONE, G_OPTION_ARG_STRING_ARRAY,
+        &arg_remaining, NULL, "validate PACKAGE" },
```
I think we should *optionally* allow validating a package against a real device. In case no device was specified, a loop device is probably a good idea still.

landerweit-phytec replied:
Added that feature with the `--device` option.

landerweit-phytec commented on tests/emmc.c:
```diff
@@ -10,6 +10,7 @@
 #include "helper.h"
 #include "pu-emmc.h"
 #include "pu-error.h"
+#include "pu-input.h"
```
Am I missing something or is this already the correct order as `i` comes after `e`.

landerweit-phytec commented on src/pu-input.c:
```diff
@@ -0,0 +1,101 @@
+
+    path = pu_path_from_uri(input->uri, prefix, error);
+    if (path == NULL) {
+        g_prefix_error(error, "Failed parsing input URI for partition: ");
```
fixed

landerweit-phytec commented on src/pu-input.c:
```diff
@@ -0,0 +1,101 @@
+
+gboolean
+pu_validate_file(PuInput *input,
+                 GError **error)
```
I moved the file existence check from `pu_set_absolute_path()` here so the new name `pu_input_validate_file()` should fit.

landerweit-phytec commented on src/pu-input.c:
```diff
@@ -0,0 +1,101 @@
+#include "pu-checksum.h"
+#include "pu-glib-compat.h"
+#include "pu-utils.h"
+#include "pu-input.h"
```
Fixed

landerweit-phytec committed:
src: Add loopdev functions

Add functions to create and remove loop devices.

Signed-off-by: Leonard Anderweit <l.anderweit@phytec.de>

landerweit-phytec committed:
src: utils: Allow copying to file destinations

Only append the source filename in case the destination is a directory.

Signed-off-by: Leonard Anderweit <l.anderweit@phytec.de>

landerweit-phytec committed:
src: main: Add validate command

The new command 'validate' validates a package either against a provided
device or a new loop device.

Signed-off-by: Leonard Anderweit <l.anderweit@phytec.de>

landerweit-phytec committed:
src: main: Add option to write package to loop device

Writing the package to the loop device is an additional validation step.
This is similar to installing the package to a real device.

Signed-off-by: Leonard Anderweit <l.anderweit@phytec.de>

landerweit-phytec committed:
src: main: Add option to build flashable image from package

The generated image can be written with dd.

Signed-off-by: Leonard Anderweit <l.anderweit@phytec.de>

landerweit-phytec committed:
tests: Add test for loopdev

Signed-off-by: Leonard Anderweit <l.anderweit@phytec.de>

mschwan-phytec commented on doc/usage.rst:
```diff
@@ -50,6 +48,15 @@ show [OPTION…] *PACKAGE*
+   -w, --write-image           Install the package to a file as an additional validation step
+   -b, --build-image=FILENAME  Build a bootable image from the package. Implies --write-image
+   --device=DEVICE             Validate against provided device
+
```
Maybe add a short sentence informing about the partup version this is available in? Something like:
```
Available since: 1.1.0
```

mschwan-phytec replied:
Very nitpicky, but could you move the "available since" string to the bottom of the `validate` paragraph (without parenthesis)? Like the GLib and other GNOME documentation does. :wink:

landerweit-phytec commented on doc/usage.rst:
```diff
@@ -50,6 +48,15 @@ show [OPTION…] *PACKAGE*
+   Validate a partup PACKAGE
+
+   -s, --skip-checksums        Skip checksum verification for all input files
+   --size=SIZE                 Size of loop device. Default is 1G
```
1G is 1GB here.

landerweit-phytec commented on src/pu-main.c:
```diff
@@ -168,6 +182,114 @@ cmd_show(PuCommandContext *context,
+
+        if (device_size < 128 * PED_MEBIBYTE_SIZE) {
+            g_set_error(error, PU_ERROR, PU_ERROR_FAILED,
+                        "File size '%s' is to small", args[1]);
```
Fixed

landerweit-phytec commented on doc/usage.rst:
```diff
@@ -50,6 +48,15 @@ show [OPTION…] *PACKAGE*
 
    -s, --size              Print the size of each file
 
+validate [OPTION…] *PACKAGE* *SIZE*
```
fixed

mschwan-phytec reviewed (changes requested):
Please rebase your changes, just in case.

landerweit-phytec commented on src/pu-utils.c:
```diff
@@ -534,3 +548,66 @@ pu_str_pre_remove(gchar *string,
+               unit_factor = PED_TERABYTE_SIZE;
+		}
+	} else {
+        unit_factor = 1;
```
Fixed

mschwan-phytec commented:
I tried running partup with the validate command and somehow my RAM got clogged up. I guess there is a memory leak or similar in your code? I checked for leftover mounted loop device but that does not seem to be the issue on first sight. When running partup I get the following generic error message:
```
10:07:13.989678 CRITICAL partup: Error writing to file: No space left on device
```
However, running anything that requires RAM then throws a similar error message, e.g. GCC:
```
../src/pu-loopdev.c:88:1: fatal error: error writing to /tmp/ccScBcPv.s: No space left on device
```
The commands I ran before are something like this:
```
sudo partup-git validate -d rauc-image.partup --size 2G -s
```
Maybe partup must return with an error code before this issue appears. Should happen if there are no checksums in the layout config and we do not provide the `-s` option.
